neuro_san/internals/run_context/langchain/tracing/neuro_san_runnable.py

```diff
@@ -0,0 +1,210 @@
+    # which is a type validator that langchain is based on which
+    # is able to use JSON schema definitions to validate fields.
+    invocation_context: InvocationContext = None
+
+    interceptor: InterceptingJournal = None
+
+    origin: List[Dict[str, Any]] = None
+
+    session_id: str = None
+
+    # Default logger
+    logger: Optional[Logger] = None
+
+    run_target: Optional[RunTarget] = None
+
+    # This guy needs to be a pydantic class and in order to have
+    # any non-pydantic non-serializable members, we need to do this.
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def __init__(
+        self,
+        **kwargs: Any,
+    ) -> None:
+        """
+        Constructor
+        """
+        # Use the run_target from the kwargs as the afunc parameter
+        # for the RunnablePassthrough.  If that doesn't exist, assume
+        # our own run_it() method is overridden and use that.
+        run_target: RunTarget = kwargs.get("run_target", self)
+        super().__init__(afunc=run_target.run_it, **kwargs)
+
+        self.logger: Logger = getLogger(self.__class__.__name__)
+
+    # pylint: disable=redefined-builtin
+    @override
+    async def ainvoke(
+        self,
+        input: Other,
+        config: RunnableConfig | None = None,
+        **kwargs: Any | None,
+    ) -> Other:
+
+        # Called by langchain infrastruture when a chain is invoked.
+        # Calling the super here means that run_it() below will be called
+        # as part of the RunnablePassthrough infrastructure.
+        _: Other = await super().ainvoke(input, config, **kwargs)
+
+        # Collect intercepted outputs to report back to the tracing infrastructure.
+        outputs: Dict[str, Any] = self.get_intercepted_outputs()
+        return outputs
+
+    async def run_it(self, inputs: Input) -> Output:
+        """
+        Transform a single input into an output.
+
+        Args:
+            inputs: The input to the `Runnable`.
+
+        Returns:
+            The output of the `Runnable`.
+        """
+        raise NotImplementedError
+
+    def prepare_runnable_config(self, session_id: str = None,
+                                callbacks: List[BaseCallbackHandler] = None,
+                                recursion_limit: int = None,
+                                use_run_name: bool = False) -> Dict[str, Any]:
+        """
+        Prepare a RunnableConfig for a Runnable invocation.  See:
+        https://python.langchain.com/api_reference/core/runnables/langchain_core.runnables.config.RunnableConfig.html
+
+        :param session_id: An id for the run
+        :param callbacks: A list of BaseCallbackHandlers to use for the run
+        :param recursion_limit: Maximum number of times a call can recurse.
+        :return: A dictionary to be used for a Runnable's invoke config.
+        """
+        agent_name: str = self.invocation_context.get_agent_name()
```

Comment:
We are now reporting the agent name as the main event of the "run_name" which is what shows up as the headline of the trace.

For the traces initiated by DataDrivenChatSession, this is all we report for the name now, and it's the agent name that shows up as the heading in something like LangSmith.

For individual agent reporting, this gets appended with the origin string.